atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -434,6 +437,96 @@ public void markTableInvolved(TableReference tableRef) {
+        Iterable<byte[]> distinctRows = getDistinctRows(rows);
+
+        hasReads = true;
+        int batchSize = getPerRowBatchSize(columnRangeSelection, Iterables.size(distinctRows));
+        BatchColumnRangeSelection perBatchSelection = BatchColumnRangeSelection.create(
+                columnRangeSelection.getStartCol(), columnRangeSelection.getEndCol(), batchSize);
+
+        Map<byte[], RowColumnRangeIterator> rawResults =
+                keyValueService.getRowsColumnRange(tableRef, distinctRows, perBatchSelection, getStartTimestamp());
+
+        return getPostFilteredSortedColumns(tableRef, columnRangeSelection, distinctRows, rawResults);
+    }
+
+    private Iterator<Map.Entry<Cell, byte[]>> getPostFilteredSortedColumns(
+            TableReference tableRef,
+            BatchColumnRangeSelection columnRangeSelection,
+            Iterable<byte[]> distinctRows,
+            Map<byte[], RowColumnRangeIterator> rawResults) {
+        Comparator<Cell> cellComparator = columnOrderThenPreserveInputRowOrder(distinctRows);
+
+        Iterator<Map.Entry<Cell, Value>> postFilterIterator = getRowColumnRangePostFilteredWithoutSorting(
+                tableRef, mergeColumnFirst(rawResults.values(), cellComparator), columnRangeSelection.getBatchHint());
+        Iterator<Map.Entry<Cell, byte[]>> remoteWrites = Iterators.transform(
+                postFilterIterator,
+                entry -> Maps.immutableEntry(entry.getKey(), entry.getValue().getContents()));
+        Iterator<Map.Entry<Cell, byte[]>> localWrites =
+                getSortedColumnsLocalWrites(tableRef, distinctRows, columnRangeSelection, cellComparator);
+        Iterator<Map.Entry<Cell, byte[]>> merged = IteratorUtils.mergeIterators(
+                localWrites,
+                remoteWrites,
+                Comparator.comparing(Map.Entry::getKey, cellComparator),
+                com.palantir.util.Pair::getLhSide);
+
+        return filterDeletedValues(merged, tableRef);
+    }
+
+    private Iterator<Map.Entry<Cell, byte[]>> getSortedColumnsLocalWrites(
+            TableReference tableRef,
+            Iterable<byte[]> rows,
+            BatchColumnRangeSelection columns,
+            Comparator<Cell> cellComparator) {
+        return mergeColumnFirst(
+                Iterables.transform(rows, row -> getLocalWritesForColumnRange(tableRef, columns, row)
+                        .entrySet()
+                        .iterator()),
+                cellComparator);
+    }
+
+    private Comparator<Cell> columnOrderThenPreserveInputRowOrder(Iterable<byte[]> rows) {
+        return Cell.columnComparator.thenComparing(
+                (Cell cell) -> ByteBuffer.wrap(cell.getRowName()),
+                Ordering.explicit(Streams.stream(rows)
+                        .map(ByteBuffer::wrap)
+                        .distinct()
+                        .collect(ImmutableList.toImmutableList())));
+    }
+
+    protected <V> Iterator<Map.Entry<Cell, V>> mergeColumnFirst(
```

Comment:
can this be private?

also, nit: mergeByComparator(...) or something, that's controlled by what you passed in (which may not be column first)